src/tests/testTokenizer.cc

```diff
@@ -248,6 +248,285 @@ TestTokenizer::testTokenizerInt64()
+        const int64_t benchmark = 04;
+        CPPUNIT_ASSERT(t.int64(rv, 8, false, 2));
+        CPPUNIT_ASSERT_EQUAL(benchmark, rv);
+        CPPUNIT_ASSERT_EQUAL(SBuf("6"), t.buf());
+    }
+
+    // auto-detected base; input: base, one invalid post-base digit, unlimited
+    {
+        int64_t rv;
+        Parser::Tokenizer t(SBuf("08"));
+        const int64_t benchmark = 00;
+        CPPUNIT_ASSERT(t.int64(rv, 8));
+        CPPUNIT_ASSERT_EQUAL(benchmark, rv);
+        CPPUNIT_ASSERT_EQUAL(SBuf("8"), t.buf());
+    }
+
+    // auto-detected base; input: base, one invalid post-base digit, limited
+    {
+        int64_t rv;
+        Parser::Tokenizer t(SBuf("08"));
+        const int64_t benchmark = 00;
+        CPPUNIT_ASSERT(t.int64(rv, 8, false, 2));
+        CPPUNIT_ASSERT_EQUAL(benchmark, rv);
+        CPPUNIT_ASSERT_EQUAL(SBuf("8"), t.buf());
+    }
+
+    // this ends four-dimensional enumeration of octal cases described earlier
+
+    // check octal base auto-detection
+    {
+        int64_t rv;
+        Parser::Tokenizer t(SBuf("0128"));
+        const int64_t benchmark = 012;
+        CPPUNIT_ASSERT(t.int64(rv, 0));
+        CPPUNIT_ASSERT_EQUAL(benchmark, rv);
+        CPPUNIT_ASSERT_EQUAL(SBuf("8"), t.buf());
+    }
+
+    // check that octal base auto-detection is not confused by repeated zeros
+    {
+        int64_t rv;
+        Parser::Tokenizer t(SBuf("00000000071"));
+        const int64_t benchmark = 00000000071;
+        CPPUNIT_ASSERT(t.int64(rv));
+        CPPUNIT_ASSERT_EQUAL(benchmark,rv);
+        CPPUNIT_ASSERT_EQUAL(SBuf(""), t.buf());
+    }
+
+    // check that forced octal base is not confused by hex prefix
+    {
+        int64_t rv;
+        Parser::Tokenizer t(SBuf("0x5"));
+        const int64_t benchmark = 0;
+        CPPUNIT_ASSERT(t.int64(rv, 8));
+        CPPUNIT_ASSERT_EQUAL(benchmark, rv);
+        CPPUNIT_ASSERT_EQUAL(SBuf("x5"), t.buf());
+    }
+
+    // autodetect decimal base in shortest valid input
+    {
+        int64_t rv;
+        Parser::Tokenizer t(SBuf("1"));
+        const int64_t benchmark = 1;
+        CPPUNIT_ASSERT(t.int64(rv));
+        CPPUNIT_ASSERT_EQUAL(benchmark,rv);
+        CPPUNIT_ASSERT(t.buf().isEmpty());
+    }
+
+    // autodetect hex base in shortest valid input
+    {
+        int64_t rv;
+        Parser::Tokenizer t(SBuf("0X1"));
+        const int64_t benchmark = 0X1;
+        CPPUNIT_ASSERT(t.int64(rv));
+        CPPUNIT_ASSERT_EQUAL(benchmark,rv);
+        CPPUNIT_ASSERT(t.buf().isEmpty());
+    }
+
+    // invalid (when autodetecting base) input matching hex base
+    {
+        int64_t rv;
+        Parser::Tokenizer t(SBuf("0x"));
+        CPPUNIT_ASSERT(!t.int64(rv));
+        CPPUNIT_ASSERT_EQUAL(SBuf("0x"), t.buf());
+    }
+
+    // invalid (when forcing hex base) input matching hex base
+    {
+        int64_t rv;
+        Parser::Tokenizer t(SBuf("0x"));
+        CPPUNIT_ASSERT(!t.int64(rv, 16));
+        CPPUNIT_ASSERT_EQUAL(SBuf("0x"), t.buf());
```

Comment:
The situation is worse than I thought. The bug discovered by Eduard is not the only previously unknown difference with strtol().

While our int64() rejects `0x` when being forced to parse hex integers (see the test case this change request is attached to), strtol() successfully parses that as ordinary zero digit (followed by "x" trailing garbage).

strtol() does not require `0x` base indicator to be present when forced to parse hex. That makes sense, of course: Our own int64() also successfully parses indicator-free "deadbeefrow" sequence (the last test case in this source file), for similar "indicator is optional when the base is known" reasons.

The outcomes differ in the `0x` input case because our int64() insists on interpreting existing `0x` prefix as a base indicator while strtol() is happy to interpret it as a case where there is no base indicator _if_ that results in a successful parsing outcome (i.e., valid hex digit "0" followed by a garbage character "x"). 

The same applies to `0x` input when int64() is told to guess the base with base=0 argument. We reject that input (see the test case immediately above the one this change request is attached to). strtol() accepts `0x` as zero followed by garbage.

In summary, unlike our Tokenizer::int64(), strtol() does not fail if its input starts with zero, _regardless_ of the valid base parameter value. Similar statements can be made for other leading digits (and non-conflicting valid base values).

Those secret differences is another set of int64() problems that we should address, but it would require serious work. Should we do that in this PR? If yes, do you agree that int64() should mimic strtol() behavior in all cases where input starts with a non-space character? CC: @kinkie

<!-- https://godbolt.org/z/aKfjzov7f -->

Reply:
> > While our int64() rejects 0x when being forced to parse hex integers, strtol() successfully parses

> It is a question whether '0x' should be parsed as a valid (0) value or rejected

I wish that were true, but the actual question is rather different! You are thinking of parsing isolated integers (for the lack of a better term). In that context, there is no doubt that "0x" is _not_ a valid integer! However, Tokenizer::int64() and strtol() do not parse isolated integers. They parse input stream prefixes. When parsing an input stream, the question is which leading portion (if any) of the given input stream is an integer. For two-character "0x" input, "0" is that integer portion (and "x" is the remaining input).

Even after int64() is fully fixed, in most (perhaps even all currently existing) int64() call contexts, the entire "0x..." input will be rejected as invalid _after_ extracting/parsing a valid integer "0" because the remaining "x..." input will not satisfy input grammar, but that caveat does _not_ affect correct int64() functionality. The method is buggy.

Reply:
> Those secret differences is another set of int64() problems that we should address. Documenting these differences is an easy way out (at least for now). Changing how int64() works would require serious work. Should we do any of that in this PR?

After giving it more thought over the weekend, I am now convinced that we should _not_ add more int64() fixes to this PR. Instead, we should address these remaining problems in followup PR(s): Unlike the "0" parsing bug fixed by this PR, the remaining problems are not trivial to fix, they do not affect known/common valid input cases, they do not block other existing PRs, and they may warrant a rather different approach to unit testing (which requires even more development work). They deserve dedicated PR(s).

@yadij, please disregard those other int64() problems and clear this PR, unblocking #1840 that depends on it.

Reply:
>  When parsing an input stream, the question is which leading portion (if any) of the given input stream is an integer.

Ok, it should be then int64() caller's decision whether, say, '0x' is an invalid integer (the end of input), or a maybe a valid one (more input is expected).

Reply:
Yes, those two are both valid caller use cases.

Reply:
Please do not fixate on bug-compatibility with `strtol()` - the `Tokenizer` is for protocol parsing. As such the limits and tolerances come from those protocol specifications, not `strtol()` behaviour.

IME, network protocols typically expect "0x" to be detected as a broken/invalid syntax. Which our `Tokenizer` needs to support detection of.  So "0x" input should indeed be detected as incomplete for base-16 or auto-detection. All other (provided non-16 bases) should detect it as `0` with trailing character(s).

Reply:
> Please do not fixate on bug-compatibility with `strtol()`

Please do not review people.

>  the `Tokenizer` is for protocol parsing. 

Tokenizer use cases are not limited to "protocol parsing". The class is already used for parsing squid.conf input, for example.

> IME, network protocols typically expect "0x" to be detected as a broken/invalid syntax. Which our `Tokenizer` needs to support detection of. So "0x" input should indeed be detected as incomplete for base-16 or auto-detection. All other (provided non-16 bases) should detect it as `0` with trailing character(s).

There are "0x is invalid input" cases you mention, but there may be other use cases as well. IMO, we should address the needs of use cases we already have (while staying as predictable/general/etc. as feasible). Doing that well is already difficult. We have [started to discuss](https://github.com/squid-cache/squid/pull/1588#discussion_r1397356692) related improvements, and I still plan to propose how to handle those needs better than we do today. All that is outside _this_ bug-fixing PR scope though, so I am not going to argue about some quoted statements that I disagree with.